src/python/deepgnn/graph_engine/snark/decoders.py

```diff
@@ -43,6 +43,191 @@ def decode(self, line: str) -> Iterator[Tuple[int, int, int, float, list]]:
+        default_edge_weight: int Same as node except for all edges.
+        default_edge_feature_types: ["dtype" or None, ...] Dtype of each feature vector.
+        default_edge_feature_lens: [[int, ...] or None, ...] Length value for each feature vector.
+
+    Init Parameters Format Example
+        Same graph as the previous example, just with defaults specified.
+        ```
+        EdgeListDecoder(
+            default_node_type=1,
+            default_node_weight=.5,
+            default_node_feature_types=["int32", "float32"],
+            default_node_feature_lens=[[3],[2]],
+            default_edge_type=0,
+            default_edge_weight=.5,
+            default_edge_feature_types=["uint8"],
+                default_edge_feature_lens=[[3, 0]],
+        )
+        ```
+        graph.csv
+        ```
+        0 -1 1 1 1 1.1 1.1
+        0 1 0 4 10 1 1 1
+        1 -1 1 1 1 1.1 1.1
+        1 0 0 4 10 1 1 1
+        ```
+    """
+
+    def __init__(
+        self,
+        default_node_type: Optional[int] = None,
+        default_edge_type: Optional[int] = None,
+        default_node_weight: Optional[float] = None,
+        default_edge_weight: Optional[float] = None,
+        default_node_feature_types: Optional[List[Optional[str]]] = None,
+        default_edge_feature_types: Optional[List[Optional[str]]] = None,
+        default_node_feature_lens: Optional[List[Optional[List[int]]]] = None,
+        default_edge_feature_lens: Optional[List[Optional[List[int]]]] = None,
+    ):
+        """Initialize the Decoder."""
+        super().__init__()
+        self.node_type = default_node_type
+        self.node_weight = default_node_weight
+        self.node_feature_types = (
+            default_node_feature_types if default_node_feature_types is not None else []
+        )
+        self.node_feature_lens = (
+            default_node_feature_lens if default_node_feature_lens is not None else []
+        )
+        self.n_node_feature = len(self.node_feature_types)
+
+        self.edge_type = default_edge_type
+        self.edge_weight = default_edge_weight
+        self.edge_feature_types = (
+            default_edge_feature_types if default_edge_feature_types is not None else []
+        )
+        self.edge_feature_lens = (
+            default_edge_feature_lens if default_edge_feature_lens is not None else []
+        )
+        self.n_edge_feature = len(self.edge_feature_types)
+
+    def decode(self, line: str) -> Iterator[Tuple[int, int, int, float, list]]:
+        """Convert text line into a node and edge iterator.
+
+        This is a generator that yields a node then its outgoing edges in order.
+        A node and its outgoing edges can be on different partitions.
+        Yield format is (node_id/src, -1/dst, type, weight, features).
+        Features being a list of dense features as ndarrays and sparse features as 2 tuples, coordinates and values.
+        """
+        if line == "":
+            return []
```

Comment:
why do we return, not yield an empty result?

Reply:
This function is used in a for loop, so here on the null case we return an empty iterator and the for loop will not even run a single loop. If we yield a None, the for loop will at least run one loop on that and if it does not break it will attempt to loop again where it'll get an error.

Reply:
I use this direclty in a for loop to stream the input, eg
for item in decoder.decode(line):
if I change this to
for item in list(decoder.decode(line)):
it will no longer stream

Reply:
I just didn't want to print a for loop in the previous example. Empty return should still work:
![image](https://user-images.githubusercontent.com/134559/183723899-0369898a-5385-4e6e-90c3-050772b6f506.png)

Reply:
Interesting, this only works if there is also a yield ni the function because normally for i in None throws an error